### ISSUE TITLE
Is maxlength attribute accessible?

### ISSUE THREAD

**Issue by Fu22yWu22y** [wet-boew/wet-boew#7144]
Hey  
i was looking up if max length attribute was accessible and came across http://stackoverflow.com/questions/24036568/how-to-make-maxlength-attribute-accessible   is there a solution in Wet for this ?  or is there one foreseen in the near future?

**Comment by LaurentGoderre** [comment]
This is more likely something that should be fixed by JAWS

**Comment by andre-gh** [comment]
Wow, that sounds really frustrating to use.  I suggest ensuring that your label mentions the max length somewhere to at-least give users a hint about what will occur.

Working around the issue directly (attempting to detect the keyboard input and signalling an error) would get very complicated very fast and would likely not function correctly in a screen reader which does support maxlength properly possibly stealing verbal focus from the input box and confusing the user.

**Comment by robwloch** [comment]
I tested with the NVDA screen reader and it just keeps announcing characters that I type even after I've reached the limit so no audible indication that what I'm typing isn't being stored in the field.

So, as a best practice, should all text input field labels have hidden text to indicate the max length in order to make it accessible?
ie.
<label for="lastname">Last name <span class="wb-inv">(maximum 20 characters)</span></label>
<input type="text" name="lastname" id="lastname" size="20" maxlength="20" class="form-control">

**Comment by andre-gh** [comment]
As a best practice max length should be available via a label for everyone If you must, the max length label may be invisible and it will solve the accessibility issue, it is a good idea to have it visible for everyone as users may plan ahead before they type the value instead of hitting the max then realizing.

**Comment by robwloch** [comment]
Thank you @andre-gh . I was thinking that since visual users would be able to see as they're typing that their input was no longer being entered that they don't need it as much. Otherwise, having every single text input field label on the page showing a max length might look awkward and cluttered. For textareas and certain text fields, I think visible makes sense but just not all. Would you agree?

**Comment by andre-gh** [comment]
As a user I find not knowing the max length ahead of time very frustrating, but from an accessibility perspective there is no issue with only the screen reader knowing about it.

**Comment by robwloch** [comment]
I agree with you for certain fields but would be overkill for all fields in my opinion:

![4github](https://cloud.githubusercontent.com/assets/5721679/11281039/c29fac96-8ec6-11e5-8390-07c2196dd66a.PNG)

**Comment by andre-gh** [comment]
Telephone and postal code have an implicit max length, however, it is a very good idea for those to put the expected format in as a placeholder text.

**Comment by andre-gh** [comment]
(or a label, I can not recall the rule off of the top of my head)  However, it is not easy to know ahead of time if you accept "XXX XXX" or "XXXXXX" or with phone "(xxx)-xxx-xxxx" etc

**Comment by robwloch** [comment]
Agreed, this is just from the WET4 examples form validation page. I just used firebug to add the character maximum to the labels to show how cluttered it might look.

http://wet-boew.github.io/wet-boew/demos/formvalid/formvalid-en.html

**Comment by hsrudnicki** [comment]
Can be this issue closed as resolved and fixed or if otherwise?
Either way, please update this issue's status, whether this issue can be closed or otherwise, thank you

**Comment by robwloch** [comment]
In researching how to make textarea input types more accessible, I came across aria-live and I'm thinking that it might be useful good practice on text fields too since screen readers do not notify users that they've reached the maxlength limit. The aria-live assertive mode can interrupt the user as they type to notify them that they've reached the limit.

Wondering what others think about this?

**Comment by hsrudnicki** [comment]
Does this remain an issue for further discussion/thoughts or otherwise?

**Comment by cybersubtype** [comment]
Problem also stems with jQuery Validation. It don't validate their is a issue with what is being type until the user changes input (leaves focus).

**Comment by cybersubtype** [comment]
Something like this along side jQuery Form Validation may work.
```javascript
function textValidation(element, max, min) {
	var length = element.value.trim().length;
	if (length > max) {
		element.value = element.value.substring(0, max);
		length = element.value.length;
		if(!$("#alert").length) {
			$(element).after('<span id="alert" aria-live="assertive"><small>Remaining characters: 0</small></span>');
		}
	} else if (length < min) {
		$(element).parent().removeClass('has-success');
	}
	if(length <= max && length >= min){
		$(element).parent().addClass('has-success');
		$(element).removeAttr('placeholder');
	}
};
```

**Comment by RobJohnston** [comment]
@cybersubtype :  Take a look at issue #6534, where the [jQuery Word and Character Counter] (http://plugins.jquery.com/word-and-character-counter/) was discussed.  Aria support was added to it in https://github.com/qwertypants/jQuery-Word-and-Character-Counter-Plugin/pull/20.

**Comment by cybersubtype** [comment]
@RobJohnston That is nice plugin but not nothing was done to implement it into WET. Be nice if it was built into the form validation on WET side. Could loop the form for all text inputs and text areas and apply the counter based on their max length.

**Comment by RobJohnston** [comment]
@cybersubtype Just thought it might be useful for what you're doing, with a lot less code to write.  I've used it on an intranet site successfully.  But I think that it's overkill for **every** form field (my 2 cents).

**Comment by cybersubtype** [comment]
Yeah it can be. Bit overkill but it can also leave someone out of the loop. Not bad wasn't a easy way for every little problem.

**Comment by RobJohnston** [comment]
There was also a discussion about the accessibility of a character counter in https://github.com/wet-boew/GCWeb/issues/968 that might be of interest.

**Comment by cybersubtype** [comment]
Thanks for the information I decide to use the jQuery Word and Character Counter I took this approach:

HTML:
```html
<div class="form-group">
	<label for="text" class="required"><span class="field-name">Text Example</span></label> 
	<input id="text" class="form-control counter-inv" type="text" name="text" required="required" data-rule-minlength="1" data-rule-maxlength="15"/>
</div>
<div class="form-group">
	<label for="title" class="required"><span class="field-name">Textarea Example</span></label> 
	<textarea  id="title" class="form-control" rows="4" cols="40" name="title" required="required" data-rule-minlength="1" data-rule-maxlength="200" ></textarea>
</div>
```

JavaScript:
```javascript
$('form input[type="text"]').each(function(){
	var max = $(this).attr('data-rule-maxlength');
	if(max) {
		var hidden = $(this).hasClass('counter-inv');
		characterCounter(this, max, hidden);
	}
});
$('form textarea').each(function(){
	var max = $(this).attr('data-rule-maxlength');
	if(max) {
		var hidden = $(this).hasClass('counter-inv');
		characterCounter(this, max, hidden);
	}
});
function characterCounter(element, max, hidden) {
    var containerClass = (hidden) ? 'wb-inv' : '';
    $(element).counter({
        goal: max,
        count: 'down',
        containerClass: containerClass
    });
};
```

**Comment by RobJohnston** [comment]
@cybersubtype : Those translations are not French... try `'caractère mots restants maximum'`.

**Comment by cybersubtype** [comment]
Thanks now new issue the french `è` is breaking the plugin.

**Comment by cybersubtype** [comment]
@RobJohnston Nevermind just a charset issue. All working thanks! :)

**Comment by LaurentGoderre** [comment]
@SBleeks @dfarough has this ever affected you?

**Comment by SBleeks** [comment]
It hasn’t effected me in the sense that max length isn’t usually a problem.
We have developed a solution here for electronic questionnaires. While most edit fields work well by default, occasionally an edit field will require a form of Max character” indication.

An example is a “comment” field. We tried using ARIA Live but even using the “polite” attribute, the UI and user experience varied depending on browser/ User agent combination. Usually the overall experience with ARIA Live was not great.
So the solution we use is 2 fold:

1.       Indicate in text the number of characters available prior to the edit field, and also state that the number of characters remaining is offered directly beneath the edit field.
This way if a blind screen reader user wants to know how many characters remain, they need only arrow down and review the number of characters which are vvisually displayed.

This approach has worked well and across all user agent/ browser conbinations.

I agree that for common edit fields such as “telephone number” and email address”, the format example is key and avoids user error.

**Comment by LaurentGoderre** [comment]
@SBleeks so would it be accurate to say that maximum number of characters are only really a problem for longer form element such as text area? Couldn't we use a character counter that have live off but user could still go back to it for information

**Comment by SBleeks** [comment]
Yes. Sounds reasonable to me. I’d need to test it first but sounds good.

I’ll give some thought to any possible relevant WCAG success criteria.
